src/profiles/profiles.service.ts

```diff
@@ -1009,6 +983,60 @@ export class ProfilesService {
       );
     return [...activeProfiles, ...archivedProfiles];
   }
+
+  public async updatePrimaryAddresses(addresses: Set<string>) {
+    for (const address of Array.from(addresses)) {
+      const profile = await this.getProfileByWallet(address);
+      if (profile?.profile && profile.consolidation.consolidation_key) {
+        const primaryAddress = await this.determinePrimaryAddress(
+          profile.consolidation.wallets,
+          profile.consolidation.consolidation_key
+        );
+        const currentPrimaryAddress = profile.profile.primary_wallet;
+        if (!areEqualAddresses(primaryAddress, currentPrimaryAddress)) {
+          await this.updateProfilePrimaryAddress(
+            profile.profile.external_id,
+            primaryAddress
+          );
+        }
+      }
+    }
+  }
+
+  private async determinePrimaryAddress(
+    wallets: { wallet: Wallet }[],
+    consolidationKey: string
+  ): Promise<string> {
+    if (wallets.length === 1) {
+      return wallets[0].wallet.address;
+    }
+
+    const delegationPrimaryAddress =
+      await getDelegationPrimaryAddressForConsolidation(consolidationKey);
+    if (delegationPrimaryAddress) {
+      return delegationPrimaryAddress;
+    }
+
+    const highestTdhAddress = await getHighestTdhAddressForConsolidationKey(
+      consolidationKey
+    );
+    if (highestTdhAddress) {
+      return highestTdhAddress;
+    }
+
+    return wallets[0].wallet.address;
```

Comment:
if no primary is set via delegation, and no wallet has highest tdh, then it just returns the first wallet in the list? 

is the list already sorted by any criteria? is it date-added, or alphabetic?